abci/types/types.go

```diff
@@ -167,3 +170,125 @@ func RespondVerifyVoteExtension(ok bool) ResponseVerifyVoteExtension {
 		Result: result,
 	}
 }
+
+func MustHashResults(r []*ExecTxResult) []byte {
+	return merkle.HashFromByteSlices(mustResultsToByteSlices(r))
+}
+
+// ProveResult returns a merkle proof of one result from the set
+func MustProveResult(r []*ExecTxResult, i int) merkle.Proof {
+	_, proofs := merkle.ProofsFromByteSlices(mustResultsToByteSlices(r))
+	return *proofs[i]
+}
+
+// deterministicExecTxResult strips non-deterministic fields from
+// ResponseDeliverTx and returns another ResponseDeliverTx.
+func deterministicExecTxResult(response *ExecTxResult) *ExecTxResult {
+	return &ExecTxResult{
+		Code:      response.Code,
+		Data:      response.Data,
+		GasWanted: response.GasWanted,
+		GasUsed:   response.GasUsed,
+	}
+}
+
+func mustResultsToByteSlices(r []*ExecTxResult) [][]byte {
+	s := make([][]byte, len(r))
+	for i, e := range r {
+		d := deterministicExecTxResult(e)
+		b, err := d.Marshal()
+		if err != nil {
+			panic(err)
+		}
+		s[i] = b
+	}
+	return s
+}
+
+func (tr *TxRecord) IsIncluded() bool {
+	return tr.Action == TxRecord_ADDED || tr.Action == TxRecord_UNMODIFIED
+}
+
+// IncludedTxs returns all of the TxRecords that are marked for inclusion in the
+// proposed block.
+func (rpp *ResponsePrepareProposal) IncludedTxs() []*TxRecord {
+	trs := []*TxRecord{}
+	for _, tr := range rpp.TxRecords {
+		if tr.IsIncluded() {
+			trs = append(trs, tr)
+		}
+	}
+	return trs
+}
+
+// RemovedTxs returns all of the TxRecords that are marked for removal from the
+// mempool.
+func (rpp *ResponsePrepareProposal) RemovedTxs() []*TxRecord {
+	trs := []*TxRecord{}
```

Comment:
preallocate?

Reply:
IMO don't preallocate here. Better still: start with nil rather than empty slice.

```go
var trs []*TxRecord
```

the expected size of the output is 0 most of the time. Preallocating is wasting space in the common case of all of these. Trust append. :)